packages/core/strapi/lib/services/entity-validator/index.js

```diff
@@ -222,10 +223,122 @@ const createValidateEntity =
         entity,
       },
       { isDraft }
-    ).required();
+    )
+      .test('relations-test', 'check that all relations exist', async function (data) {
+        try {
+          await checkRelationsExist(buildRelationsStore({ uid: model.uid, data }));
+        } catch (e) {
+          return this.createError({
+            path: this.path,
+            message: e.message,
+          });
+        }
+        return true;
+      })
+      .required();
+
     return validateYupSchema(validator, { strict: false, abortEarly: false })(data);
   };
 
+/**
+ * Builds an object containing all the media and relations being associated with an entity
+ * @param {String} uid of the model
+ * @param {Object} data
+ * @param {Object} relationsStore to be updated and returned
+ * @returns
+ */
+const buildRelationsStore = ({ uid, data, relationsStore = {} }) => {
+  if (isEmpty(data)) {
+    return relationsStore;
+  }
+
+  const currentModel = strapi.getModel(uid);
+
+  Object.keys(currentModel.attributes).forEach((attributeName) => {
+    const attribute = currentModel.attributes[attributeName];
+
+    const value = data[attributeName];
+    if (isNil(value)) {
+      return;
+    }
+
+    switch (attribute.type) {
+      case 'relation':
+      case 'media': {
+        if (attribute.relation === 'morphToMany' || attribute.relation === 'morphToOne') {
+          // TODO: handle polymorphic relations
+          break;
+        }
+
+        const target = attribute.type === 'media' ? 'plugin::upload.file' : attribute.target;
+        // As there are multiple formats supported for associating relations
+        // with an entity, the value here can be an: array, object or number.
+        let source;
+        if (Array.isArray(value)) {
+          source = value;
+        } else if (isObject(value)) {
+          source = value?.connect ?? value?.set ?? [];
+        } else {
+          source = castArray(value);
+        }
+        const idArray = source.map((v) => ({ id: v.id || v }));
+
+        // Update the relationStore to keep track of all associations being made
+        // with relations and media.
+        relationsStore[target] = relationsStore[target] || [];
+        relationsStore[target].push(...idArray);
```

Comment:
I'm not comfortable with the way we always modify references in functions. Are we planning to use pure functions or is that something we don't want to do?